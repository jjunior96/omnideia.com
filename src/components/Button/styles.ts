import { darken } from 'polished';
import styled, { css, DefaultTheme } from 'styled-components';

import { ButtonProps } from '.';

const containerModifiers = {
  small: (theme: DefaultTheme) => css`
    height: 2.4rem;

    font-size: ${theme.font.sizes.xxsmall};
  `,

  medium: (theme: DefaultTheme) => css`
    height: 3.2rem;

    font-size: ${theme.font.sizes.xsmall};
  `,

  large: (theme: DefaultTheme) => css`
    height: 5.6rem;

    font-size: ${theme.font.sizes.xsmall};
  `,

  minimal: (theme: DefaultTheme) => css`
    background: none;

    color: ${theme.colors.green};

    font-weight: ${theme.font.normal};

    transition: color 0.2s;

    &:hover {
      color: ${darken(0.1, theme.colors.green)};
    }
  `
};

export const Container = styled.button<ButtonProps>`
  ${({ theme, color, size, minimal }) => css`
    width: 100%;

    color: ${theme.colors.white};
    background-color: ${theme.colors[color!]};

    display: flex;
    justify-content: center;
    align-items: center;

    border-radius: ${theme.border.radius};
    border: 0;

    padding: 0 ${theme.spacings.xsmall};

    font-weight: ${theme.font.bold};
    font-size: ${theme.font.sizes.xsmall};

    transition: background-color 0.3s, box-shadow 0.2s;

    outline: none;

    &:focus {
      box-shadow: 0 0 0.5rem ${theme.colors.green};
    }

    ${!minimal &&
    css`
      &:hover {
        background-color: ${darken(0.05, theme.colors[color!])};

        box-shadow: ${theme.box.shadow};
      }
    `}

    > svg {
      margin-right: ${theme.spacings.xxsmall};
    }

    ${!!size && containerModifiers[size](theme)};

    ${!!minimal && containerModifiers.minimal(theme)};
  `}
`;
