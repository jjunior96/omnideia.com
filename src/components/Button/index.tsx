import React, {
  AnchorHTMLAttributes,
  ButtonHTMLAttributes,
  forwardRef
} from 'react';

import * as S from './styles';

type ButtonTypes =
  | AnchorHTMLAttributes<HTMLAnchorElement>
  | ButtonHTMLAttributes<HTMLButtonElement>;

type Props = {
  size?: 'small' | 'medium' | 'large';
  color?: 'primary' | 'secundary';
  children: React.ReactNode;
  minimal?: boolean;
  icon?: React.ReactNode;
  positionIcon?: 'right' | 'left';
  as?: React.ElementType;
};

export type ButtonProps = ButtonTypes & Props;

const Button: React.ForwardRefRenderFunction<ButtonProps, ButtonProps> = (
  {
    children,
    size = 'large',
    color = 'primary',
    minimal = false,
    icon,
    positionIcon = 'left',
    as,
    ...rest
  },
  ref
) => {
  return (
    <S.Container
      as={as}
      size={size}
      positionIcon={positionIcon}
      color={color}
      minimal={minimal}
      ref={ref}
      {...rest}
    >
      <S.Content>
        {!!icon && (
          <S.IconContainer positionIcon={positionIcon}>{icon}</S.IconContainer>
        )}
        {children}
      </S.Content>
    </S.Container>
  );
};

export default forwardRef(Button);
