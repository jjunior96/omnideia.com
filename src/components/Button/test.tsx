import { screen } from '@testing-library/react';

import { renderWithTheme } from 'utils/tests/helpers';

import theme from 'styles/theme';

import Button from '.';

describe('<Button />', () => {
  it('should render a button default', () => {
    const { container } = renderWithTheme(<Button>Enter</Button>);

    expect(screen.getByText(/Enter/i)).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('should render a button with size small', () => {
    renderWithTheme(<Button size="small">Enter</Button>);

    expect(screen.getByRole('button', { name: /Enter/i })).toHaveStyle({
      height: '2.4rem'
    });
  });

  it('should render a button with size large', () => {
    renderWithTheme(<Button size="large">Enter</Button>);

    expect(screen.getByRole('button', { name: /Enter/i })).toHaveStyle({
      height: '5.6rem'
    });
  });

  it('should render a button with color primary', () => {
    renderWithTheme(<Button color="primary">Enter</Button>);

    expect(screen.getByRole('button', { name: /Enter/i })).toHaveStyle({
      'background-color': theme.colors.primary
    });
  });

  it('should render a button with color secundary', () => {
    renderWithTheme(<Button color="secundary">Enter</Button>);

    expect(screen.getByRole('button', { name: /Enter/i })).toHaveStyle({
      'background-color': theme.colors.secundary
    });
  });

  it('should render a minimal version', () => {
    renderWithTheme(<Button minimal>Enter</Button>);

    expect(screen.getByRole('button', { name: /Enter/i })).toHaveStyle({
      background: 'none',
      color: theme.colors.green
    });
  });

  it('should render a button as a link (<a/>)', () => {
    renderWithTheme(
      <Button color="primary" as="a" href="/">
        Enter
      </Button>
    );

    expect(screen.getByRole('link', { name: /Enter/i })).toHaveAttribute(
      'href',
      '/'
    );
  });
});
