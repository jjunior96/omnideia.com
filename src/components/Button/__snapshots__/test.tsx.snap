// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render a button default 1`] = `
.c0 {
  width: 100%;
  color: #FAFAFA;
  background-color: #5855E9;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 0.8rem;
  border: 0;
  padding: 0 1.6rem;
  padding-left: 2.4rem;
  font-weight: 600;
  font-size: 1.6rem;
  -webkit-transition: background-color 0.3s,box-shadow 0.2s,outline 0.1s;
  transition: background-color 0.3s,box-shadow 0.2s,outline 0.1s;
  outline: none;
  height: 5.6rem;
  font-size: 1.6rem;
}

.c0:focus {
  outline: 3px solid #b1aff5;
}

.c0:hover {
  background-color: #423ee6;
  box-shadow: 0 8px 8px rgba(0,0,0,0.1);
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div>
  <button
    class="c0"
    color="primary"
  >
    <div
      class="c1"
    >
      Enter
    </div>
  </button>
</div>
`;
